src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,174 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node newNode = new Node(value);
+        if (isEmpty()) {
+            head = tail = newNode;
+        } else {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node nextNode = getNode(index);
+        Node prevNode = nextNode.prev;
+        Node node = new Node(prevNode, value, nextNode);
+        if (isHead(index)) {
+            head = node;
+        } else if (isTail(index - 1)) {
+            tail = node;
+        } else {
+            prevNode.next = node;
+            nextNode.prev = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T e : list) {
+            add(e);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkExistedIndex(index);
+        Node node = getNode(index);
+        return (T) node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkExistedIndex(index);
+        Node node = getNode(index);
+        T oldValue = (T) node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkExistedIndex(index);
+        Node removedNode = getNode(index);
+        T value = (T) removedNode.value;
+        if (isHead(index) && isTail(index)) {
+            head = tail = null;
+        } else if (isHead(index)) {
+            Node newHead = head.next;
+            head = newHead;
+        } else if (isTail(index)) {
+            Node newTail = tail.prev;
+            newTail.next = null;
+            tail = newTail;
+        } else {
+            Node prev = removedNode.prev;
+            Node next = removedNode.next;
+            prev.next = next;
+            next.prev = prev;
+        }
+        size--;
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node current = head;
```

Comment:
```suggestion
        int index = findIndexByValue(object);
        if (index == -1) {
            return false;
        } else {
            unlink(findNodeByIndex(index));
        }
        return true;
```

Create separate method
```
private int findIndexByValue(T value) {
        Node<T> node = head;
        for (int i = 0; i < size; i++) {
            if (value == node.item || value != null && value.equals(node.item)) {
                return i;
            }
            node = node.next;
        }
        return -1;
    }
```

Reply:
\+